background/redux-slices/earn.ts

```diff
@@ -0,0 +1,327 @@
+  name?: string
+  version?: string
+  chainId?: number
+  verifyingContract?: HexString
+}
+
+export type PermitRequest = {
+  account: HexString
+  liquidityTokenAddress: HexString
+  liquidityAmount: BigNumber
+  nonce: BigNumber
+  deadline: BigNumber
+  spender: HexString
+}
+
+export type SignTypedDataRequest = {
+  account: string
+  typedData: EIP712TypedData
+}
+
+// once testnet contracts are deployed we should replace this
+const APPROVAL_TARGET_CONTRACT_ADDRESS =
+  "0x68b3465833fb72A70ecDF485E0e4C7bD8665Fc45" // currently: swap router
+
+export const vaultDeposit = createBackgroundAsyncThunk(
+  "signing/vaultAndDeposit",
+  async (
+    {
+      vaultContractAddress,
+      amount,
+    }: {
+      tokenContractAddress: HexString
+      vaultContractAddress: HexString
+      amount: BigInt
+    },
+    { getState }
+  ) => {
+    const provider = getProvider()
+    const signer = provider.getSigner()
+    const signerAddress = await getSignerAddress()
+
+    const state = getState()
+    const { earn } = state as { earn: EarnState }
+
+    const vaultContract = await getContract(vaultContractAddress, VAULT_ABI)
+
+    const depositTransactionData =
+      await vaultContract.populateTransaction.depositWithApprovalTarget(
+        amount,
+        signerAddress,
+        signerAddress,
+        amount,
+        (await provider.getBlock(provider.getBlockNumber())).timestamp +
+          3 * HOUR,
+        earn.signature.r,
+        earn.signature.s,
+        earn.signature.v
+      )
+    signer.sendTransaction(depositTransactionData)
+  }
+)
+
+export const vaultWithdraw = createBackgroundAsyncThunk(
+  "earn/vaultWithdraw",
+  async ({
+    vaultContractAddress,
+    amount,
+  }: {
+    vaultContractAddress: HexString
+    amount: BigNumber
+  }) => {
+    const provider = getProvider()
+    const signer = provider.getSigner()
+
+    const vaultContract = new ethers.Contract(
+      vaultContractAddress,
+      VAULT_ABI,
+      signer
+    )
+    const signedWithdrawTransaction = await signer.signTransaction(
+      await vaultContract.functions["withdraw(uint256)"](amount)
+    )
+
+    provider.sendTransaction(signedWithdrawTransaction)
+  }
+)
+
+export const getRewards = createBackgroundAsyncThunk(
+  "earn/getRewards",
+  async (vaultContractAddress: HexString) => {
+    const provider = getProvider()
+    const signer = provider.getSigner()
+
+    const vaultContract = new ethers.Contract(
+      vaultContractAddress,
+      VAULT_ABI,
+      signer
+    )
+    const signedGetRewardsTx = await signer.signTransaction(
+      await vaultContract.functions.getReward()
+    )
+
+    provider.sendTransaction(signedGetRewardsTx)
+  }
+)
+
+const earnSlice = createSlice({
+  name: "earn",
+  initialState,
+  reducers: {
+    saveSignature: (
+      state,
+      { payload: { r, s, v } }: { payload: { r: string; s: string; v: number } }
+    ) => ({
+      ...state,
+      signature: {
+        r,
+        s,
+        v,
+      },
```

Comment:
_Really_ minor: it looks better on a single line (and Prettier should allow it)
```suggestion
      signature: { r, s, v },
```

Reply:
It does look better! :)